core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -1246,6 +1333,11 @@ public SELF withTmpFs(Map<String, String> mapping) {
         return self();
     }
 
+    public SELF withReuse(boolean reusable) {
```

Comment:
Re this bit of the API, I'm afraid I still really would like this to not be a boolean, so that we have room for expansion. e.g. to have an enum or interface to allow (pseudo-code-ish):

```
.withReuse( Reuse.reuseContainers() )
```

but also allow, for the sake of #1713 (and the [hilariously old draft implementation](https://github.com/testcontainers/testcontainers-java/commit/446b2feaacae5711a536d7ec6ad361b0a2a55732)):

```
.withReuse ( Reuse.reuseImagesAfterInitialization() ) // definitely needs a better name
```

I think we could get the second style of reuse done very quickly for JDBC-based containers.

With the reduced scope of _container reuse_ in this PR, do you think this buys us enough cognitive space to fit this flexibility in?

Reply:
While I understand why we would want to add it, here my 2c why I think we should not do it as part of the first step:
1. JDBC URL can only have boolean (or lambda reference, but that's tricky)
2. `withReuse(Boolean)` is easy to extend later with an overloaded `withReuse(ReuseStrategy)`
3. The image caching is another great optimization, but I don't feel that it is the same as the container reusing - we will need to start a new container every time (which is also CI friendly, unlike this PR), which kinda makes it a sugar for the image + capturing an image after the start of container.
4. The reuse is currently has to be enabled with the property. Having different strategies may make it harder because I assume it will be per-strategy enablement

The implementation details of the reuse also make it not easy to have strategies for it, and we may never have more than one (given №3)

Reply:
OK, I'm convinced 😄 

I'll add a 5th reason: while I'd love to do it, we may never get around to image reuse. To be pragmatic, I don't want to hold up this almost feature for a someday-maybe feature.